packages/react/popper/src/Popper.tsx

```diff
@@ -18,17 +18,78 @@ import type { Measurable } from '@radix-ui/rect';
 const POPPER_NAME = 'Popper';
 
 type PopperContextValue = {
+  anchorRef: React.MutableRefObject<Measurable | null>;
+};
+const [PopperProvider, usePopperContext] = createContext<PopperContextValue>(POPPER_NAME);
+
+const Popper: React.FC = ({ children }) => {
+  const anchorRef = React.useRef<Measurable | null>(null);
+  return <PopperProvider anchorRef={anchorRef}>{children}</PopperProvider>;
+};
+
+Popper.displayName = POPPER_NAME;
+
+/* -------------------------------------------------------------------------------------------------
+ * PopperAnchor
+ * -----------------------------------------------------------------------------------------------*/
+
+const ANCHOR_NAME = 'PopperAnchor';
+
+type PopperAnchorOwnProps = Polymorphic.Merge<
+  Polymorphic.OwnProps<typeof Primitive>,
+  { virtualRef?: React.RefObject<Measurable> }
+>;
+type PopperAnchorPrimitive = Polymorphic.ForwardRefComponent<
+  Polymorphic.IntrinsicElement<typeof Primitive>,
+  PopperAnchorOwnProps
+>;
+
+const PopperAnchor = React.forwardRef((props, forwardedRef) => {
+  const { virtualRef, children, ...anchorProps } = props;
+  const context = usePopperContext(ANCHOR_NAME);
+  const ref = React.useRef<React.ElementRef<typeof Primitive>>(null);
+  const composedRefs = useComposedRefs(
+    forwardedRef,
+    ref,
+    context.anchorRef as React.MutableRefObject<React.ElementRef<typeof Primitive>>
+  );
+
+  React.useEffect(() => {
+    if (virtualRef?.current) {
+      context.anchorRef.current = virtualRef.current;
+    }
+  });
+
+  return virtualRef ? (
+    <>{children}</>
```

Comment:
Do we need the fragment here?

Reply:
Or actually, do we even care about children of the anchor if it's a virtual ref? I guess it's useful but when I read this I think, well if I have a virtual ref, I don't have an element, so I should be able to use the `Anchor` part as a sibling, I don't need to wrap some children.

Reply:
Yep, happy to do that, I did it that way too initially. Just means that `ContextMenu` would be like this instead:

```jsx
<>
  <Anchor virtualRef={} />
  <Primitive />
</>
```

I quite liked that it wrapped it though because I had meant to raise whether you think it would be nice to change `virtualRef` prop to something like:

```jsx
<Anchor toPointer>
  <Primitive />
</Anchor>
```

And then it would bind the `mousemove` stuff for you on the `Primitive`. That seems like a nice API but wasn't sure if this is the only real use-case for `virtualRef` atm?

Reply:
Ah scrap that, I remember why I didn't raise that in the end... because `ContextMenu` is `onContextMenu`... 😅 

Yeah, okay, nevermind... I'll remove `children`.